[miele] Temperature channel improvements: UoM, categories and descriptions

jlaur opened the pull request:
This PR improves channels provided by the binding in different ways.

### Temperatures
- Changed type from **Number** to **Number:Temperature**.
- Now supports Unit of Measurement.
- Fixed descriptions for specific appliances. For example, previously target temperature channels were described as "Target temperature to be reached by the oven", but in fact applied also to fridge and washing machine.
- Special value 32768 °C is now correctly translated into **Undefined**.
- Added category **Temperature** for default icons.

### Spinning speed
For washing machines spinning speed was multiplied by 10. Probably mislead by description in response JSON:

```
                {
                    "Name": "spinningSpeed",
                    "Value": 1600,
                    "Polling": 0,
                    "Metadata": {
                        "Filter": "true",
                        "description": "Spinning speed divided by 10!, 0x00 = without spinning, 0xFF = rinse hold.",
                        "LocalizedID": "Spin Speed",
                        "LocalizedValue": "1600",
                        "access": "RE"
                    }
                },
```

16.000 is clearly not correct number of rotations per minute.

### Time
Added category **Time** for the following channels:
- start
- duration
- elapsed
- finish

**Elapsed** and **remaining** time channels are no longer marked as advanced.

### Next steps
See #11317. Should be linked by first commit, but isn't.

Added myself as binding maintainer, see #11321.

lolodomo commented on bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/ExtendedDeviceStateUtil.java:
```diff
@@ -46,4 +52,15 @@ public static String bytesToHex(byte[] bytes) {
+     * Convert string to Number:Temperature state with unit Celcius
+     */
+    public static State getTemperatureState(String s) throws NumberFormatException {
+        if (s.equals(TEMPERATURE_UNDEFINED)) {
```
Can you please reverse ? You should certainly have a SAT warning on this line.

jlaur replied:
Can you elaborate, I don't understand?

lolodomo replied:
Constant at left: TEMPERATURE_UNDEFINED.equals(s)

Maybe the SAT check is more clever than I imagine and only alerts when the variable can be null ?

Skinah replied:
go to the bindings \target\code-analysis folder after you compile and read the report for warnings.

Yes I believe it is more clever as I made a suggestion to someone else and got told it did not apply.

jlaur replied:
Thanks, fixed.

lolodomo commented on bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/FridgeChannelSelector.java:
```diff
@@ -142,6 +143,15 @@ public State getState(String s) {
+    public State getTemperatureState(String s) {
+        try {
+            return ExtendedDeviceStateUtil.getTemperatureState(s);
+        } catch (Exception e) {
```
Please avoid catching a so generic exception.
Isn't `NumberFormatException ` enough ?

jlaur replied:
Yes, fixed.

lolodomo commented on bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/FridgeChannelSelector.java:
```diff
@@ -142,6 +143,15 @@ public State getState(String s) {
+        try {
+            return ExtendedDeviceStateUtil.getTemperatureState(s);
+        } catch (Exception e) {
+            logger.error("An exception occurred while converting '{}' into a State", s);
```
WARN level is certainly more adapted.

jlaur replied:
Agreed, fixed.

lolodomo commented on bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/FridgeFreezerChannelSelector.java:
```diff
@@ -160,6 +161,15 @@ public State getState(String s) {
+    public State getTemperatureState(String s) {
+        try {
+            return ExtendedDeviceStateUtil.getTemperatureState(s);
+        } catch (Exception e) {
```
Please avoid catching a so generic exception.
Isn't `NumberFormatException ` enough ?

jlaur replied:
Yes, fixed.

lolodomo commented on bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/FridgeFreezerChannelSelector.java:
```diff
@@ -160,6 +161,15 @@ public State getState(String s) {
+        try {
+            return ExtendedDeviceStateUtil.getTemperatureState(s);
+        } catch (Exception e) {
+            logger.error("An exception occurred while converting '{}' into a State", s);
```
WARN level is certainly more adapted.

jlaur replied:
Agreed, fixed.

lolodomo commented on bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/FridgeChannelSelector.java:
```diff
@@ -142,6 +143,15 @@ public State getState(String s) {
+            return ExtendedDeviceStateUtil.getTemperatureState(s);
+        } catch (Exception e) {
+            logger.error("An exception occurred while converting '{}' into a State", s);
+            return null;
```
Rather than null, wouldn't it be better to return UndefType.UNDEF ?
In case there is a valid reason to return null, please add `@Nullable` before State in the method declaration.

jlaur replied:
Good idea, fixed.

lolodomo commented on bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/FridgeFreezerChannelSelector.java:
```diff
@@ -160,6 +161,15 @@ public State getState(String s) {
+            return ExtendedDeviceStateUtil.getTemperatureState(s);
+        } catch (Exception e) {
+            logger.error("An exception occurred while converting '{}' into a State", s);
+            return null;
```
Rather than null, wouldn't it be better to return UndefType.UNDEF ?
In case there is a valid reason to return null, please add `@Nullable` before State in the method declaration.

jlaur replied:
Good idea, fixed.

lolodomo commented on bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/OvenChannelSelector.java:
```diff
@@ -216,6 +217,15 @@ public State getState(String s) {
+    public State getTemperatureState(String s) {
+        try {
+            return ExtendedDeviceStateUtil.getTemperatureState(s);
+        } catch (Exception e) {
```
Please avoid catching a so generic exception.
Isn't `NumberFormatException ` enough ?

jlaur replied:
Yes, fixed.

lolodomo commented on bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/OvenChannelSelector.java:
```diff
@@ -216,6 +217,15 @@ public State getState(String s) {
+        try {
+            return ExtendedDeviceStateUtil.getTemperatureState(s);
+        } catch (Exception e) {
+            logger.error("An exception occurred while converting '{}' into a State", s);
```
WARN level is certainly more adapted.

jlaur replied:
Agreed, fixed.

lolodomo commented on bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/OvenChannelSelector.java:
```diff
@@ -216,6 +217,15 @@ public State getState(String s) {
+            return ExtendedDeviceStateUtil.getTemperatureState(s);
+        } catch (Exception e) {
+            logger.error("An exception occurred while converting '{}' into a State", s);
+            return null;
```
Rather than null, wouldn't it be better to return UndefType.UNDEF ?
In case there is a valid reason to return null, please add `@Nullable` before State in the method declaration.

jlaur replied:
Good idea, fixed.

lolodomo commented on bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/WashingMachineChannelSelector.java:
```diff
@@ -223,6 +223,15 @@ public State getState(String s) {
+            return ExtendedDeviceStateUtil.getTemperatureState(s);
+        } catch (Exception e) {
+            logger.error("An exception occurred while converting '{}' into a State", s);
+            return null;
```
Rather than null, wouldn't it be better to return UndefType.UNDEF ?
In case there is a valid reason to return null, please add `@Nullable` before State in the method declaration.

jlaur replied:
Good idea, fixed.

lolodomo commented on bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/WashingMachineChannelSelector.java:
```diff
@@ -223,6 +223,15 @@ public State getState(String s) {
+        try {
+            return ExtendedDeviceStateUtil.getTemperatureState(s);
+        } catch (Exception e) {
+            logger.error("An exception occurred while converting '{}' into a State", s);
```
WARN level is certainly more adapted.

jlaur replied:
Agreed, fixed.

lolodomo commented on bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/WashingMachineChannelSelector.java:
```diff
@@ -223,6 +223,15 @@ public State getState(String s) {
+    public State getTemperatureState(String s) {
+        try {
+            return ExtendedDeviceStateUtil.getTemperatureState(s);
+        } catch (Exception e) {
```
Please avoid catching a so generic exception.
Isn't `NumberFormatException ` enough ?

jlaur replied:
Yes, fixed.

lolodomo commented on bundles/org.openhab.binding.miele/src/main/resources/OH-INF/thing/channeltypes.xml:
```diff
@@ -86,98 +90,50 @@
-
-	<channel-type id="temp" advanced="true">
-		<item-type>Number</item-type>
+	<channel-type id="temperature">
```
Defining so generic channel type is not recommended.
At least, you could have `temperature` and `TargetTemperature` which are two very different notions.

jlaur replied:
I considered it, and it would make sense if target temperature was actually settable. Unfortunately it's not, so decided to go with shared definition as all properties are identical, so defining two different types seems arbitrary and redundant. Arbitrary because previously we had a channel type for each individual channel. Having two is better than having nine, but still - why two?

lolodomo replied:
@kaikreuzer : please comment, I remember you told me to not do that (defining ulmtra generic chan,nel types like "temperature") a very long time ago. Is it still the case ?

jlaur replied:
You convinced me to split into current and target for reuse of same label - fixed. Still interested in comments about finding right balance between "too generic" and "too redundant". :-)

lolodomo commented on bundles/org.openhab.binding.miele/src/main/resources/OH-INF/thing/channeltypes.xml:
```diff
@@ -212,14 +168,14 @@
 	</channel-type>
 
-	<channel-type id="powerConsumption" advanced="false">
+	<channel-type id="powerConsumption">
```
You suppressed "advanced" to all channel ?
When you have many channels on a thing, this is reasonable to have some of them being advanced.
In general, you don't set a channel as advanced when you are sure that any user will use/link it.

jlaur replied:
No, only for remaining/elapsed time as I think these channels are quite common. almost the primary/most important ones. For the rest I just removed explicit attribute as **advanced=false** is default. This makes it more readable IMHO.

lolodomo replied:
Oh you're right, I did not take care that advanced attribute was set to false.
In this case, this is OK for me.

lolodomo commented on bundles/org.openhab.binding.miele/src/main/resources/OH-INF/thing/fridge.xml:
```diff
@@ -16,8 +16,14 @@
 			<channel id="supercool" typeId="supercool"/>
-			<channel id="current" typeId="current"/>
-			<channel id="target" typeId="target"/>
+			<channel id="current" typeId="temperature">
+				<label>Current Temperature</label>
+				<description>Current temperature in the fridge</description>
+			</channel>
+			<channel id="target" typeId="temperature">
+				<label>Target Temperature</label>
+				<description>Target temperature to be reached by the fridge</description>
+			</channel>
```
As already mentioned, I don't thing this is a good idea to have the same channel type. Having both will then lead to no need to redefine the label and description here.

jlaur replied:
The description is needed as it's tailored for each appliance type. But you are right about label, so will reconsider.

jlaur replied:
You convinced me to split into current and target for reuse of same label - fixed.

lolodomo commented on bundles/org.openhab.binding.miele/src/main/resources/OH-INF/thing/fridgefreezer.xml:
```diff
@@ -19,10 +19,22 @@
-			<channel id="freezertarget" typeId="freezertarget"/>
-			<channel id="fridgecurrent" typeId="fridgecurrent"/>
-			<channel id="fridgetarget" typeId="fridgetarget"/>
+			<channel id="freezercurrent" typeId="temperature">
+				<label>Current Temperature</label>
+				<description>Current temperature in the freezer compartment</description>
+			</channel>
+			<channel id="freezertarget" typeId="temperature">
+				<label>Target Temperature</label>
+				<description>Target temperature to be reached by the freezer compartment</description>
+			</channel>
+			<channel id="fridgecurrent" typeId="temperature">
+				<label>Current Temperature</label>
+				<description>Current temperature in the fridge compartment</description>
+			</channel>
+			<channel id="fridgetarget" typeId="temperature">
+				<label>Target Temperature</label>
+				<description>Target temperature to be reached by the fridge compartment</description>
+			</channel>
```
Same remark here.
But here, it makes sense to have only one common "targetTemperature channel type for 2 different channels."

jlaur replied:
You convinced me to split into current and target for reuse of same label - fixed.

lolodomo commented on bundles/org.openhab.binding.miele/src/main/resources/OH-INF/thing/oven.xml:
```diff
@@ -22,10 +22,22 @@
-			<channel id="measured" typeId="measured"/>
-			<channel id="temp1" typeId="temp"/>
-			<channel id="temp2" typeId="temp"/>
+			<channel id="target" typeId="temperature">
+				<label>Target Temperature</label>
+				<description>Target temperature to be reached by the oven</description>
+			</channel>
+			<channel id="measured" typeId="temperature">
+				<label>Measured Temperature</label>
+				<description>Actual measured temperature in the oven</description>
+			</channel>
+			<channel id="temp1" typeId="temperature">
+				<label>Temperature</label>
+				<description>Program temperature in the oven</description>
+			</channel>
+			<channel id="temp2" typeId="temperature">
+				<label>Temperature</label>
+				<description>Program temperature in the oven</description>
+			</channel>
```
Same remark

jlaur replied:
You convinced me to split into current and target for reuse of same label - fixed.

Skinah commented:
When this is merged can you create an Alert for it in this file? See other bindings that have Alerts for changing to use UOM...

https://github.com/openhab/openhab-distro/blob/34c684842beec47b1e51febd5ffb670f537504e7/distributions/openhab/src/main/resources/bin/update.lst

lolodomo commented on bundles/org.openhab.binding.miele/src/main/resources/OH-INF/thing/oven.xml:
```diff
@@ -22,10 +22,21 @@
+				<label>Measured Temperature</label>
+				<description>Actual measured temperature in the oven</description>
+			</channel>
+			<channel id="temp1" typeId="temperature">
+				<label>Temperature</label>
+				<description>Program temperature in the oven</description>
+			</channel>
+			<channel id="temp2" typeId="temperature">
+				<label>Temperature</label>
+				<description>Program temperature in the oven</description>
+			</channel>
```
Maybe use "Program Temperature" as label ?
And now way to give a hint of the difference between these 2 temperatures ?

jlaur replied:
Existing labels, not changed. Unfortunately I don't know what they are as my oven doesn't support them. I have changed to "Program Temperature 1" and "Program Temperature 2".

lolodomo reviewed (commented):
A last comment/remark

lolodomo commented:
@Skinah : why did you add the WIP label ?
I think there is no WIP, @jlaur is just considering the review comments...

Skinah commented:
Sorry label removed, saw `draft` and assumed it meant the same thing.

lolodomo reviewed (approved):
LGTM. Thanks

lolodomo commented:
Will merge if the build is green.

jlaur commented:
@Skinah - the "(potentially) not backward compatible", is that because of the change from Number to Number:Temperature? It shouldn't break compatibility, but some custom formatting might change of course.

jlaur commented:
@lolodomo - can you link to #11317? I think there is some problem with auto-linking based on commit messages when PR is born as draft.

Skinah commented:
@jlaur Yes from what I understand (happy for someone to tell me i am wrong) when you make this change from Number to Number:Temperature you will need to do a delete of the thing and recreate items. This is why I tagged it and also suggested the Alert be added to the file above. For example...

ALERT;TP-Link Smart Home Binding: The energy and rssi channels are now using Units of Measurements. Items must be adapted and the things must be recreated.

Thanks for updating the binding and adding UOM.

jlaur commented:
> @jlaur Yes from what I understand (happy for someone to tell me i am wrong) when you make this change from Number to Number:Temperature you will need to do a delete of the thing and recreate items. 

For sure you don't need to delete and recreate the thing. I will perform a new test of item created from UI. For me everything worked right out of the box, and I could simply change the type of my item in .items file for proper formatting in sitemaps etc.

jlaur commented:
> For sure you don't need to delete and recreate the thing. I will perform a new test of item created from UI. For me everything worked right out of the box, and I could simply change the type of my item in .items file for proper formatting in sitemaps etc.

@Skinah - I went back to version from main branch and created a new item from UI:
![image](https://user-images.githubusercontent.com/19519842/135716127-b5bbc5bd-9965-43fe-9eca-e2bddde40bde.png)
(same formatting in sitemap with no custom formatting: `Text item=Fridge_TargetTemperature`)

Then installed version from this branch, and changed target temperature to different value to see if still updated correctly:
![image](https://user-images.githubusercontent.com/19519842/135716363-37d27ab5-d986-4df6-9089-97183da15a7d.png)

Notice that item type doesn't match anymore, but without causing any problems. Last step, changed type of item in UI. Result:
![image](https://user-images.githubusercontent.com/19519842/135716453-b959492b-a8ca-4bcf-a024-7123eb392fd5.png)

So you don't need to recreate either things or items, you can simply update items to have UoM formatting. So this should be fully backwards compatible.
